### PR TITLE
Constants instead of literals

### PR DESCRIPTION
**Pull request by HaKr**
Nathan,

First of all, thank you for the very helpful course on Skillshare. I was pleasantly surprised by the recreation of this arcade classic to showcase various Rust topics.

You invited your audience to participate, so here my humble attempt.

### Constants instead of in-code literals
You already used the constant values NUM_COLS and NUM_ROWS. To me, using constant values is another way to prevent subtle logical errors. Especially in ranges and when checking boundaries, consequent usage of constants may prevent failures when one of the major values changes.

### Terminal teardown
I was unaware of the alternate screen, but I really like what it means for the game experience. However, the setup of the alternate screen has to be undone as well. This is done at the end of the main function. This calls for a dedicated struct, which can showcase the implementation of the Drop trait.

### The two faces of the Invaders
The way you gave the invaders two different faces, to mimic the atmosphere of the original game, really brought a smile to my face. Boldly, I wanted to extend this idea, by having each row different faces for the invaders. To me, this combined good with the use of constants and also have a dependent iterator to determine the faces to use.

I am curious what you think of it.

### CODE REVIEW

File: src/invaders.rs [hunk 1]
```diff
@@ -9,6 +9,8 @@ pub struct Invader {
     pub x: usize,
     pub y: usize,
     points: u16,
+    face1: &'static char,
+    face2: &'static char,
```
Comment (CleanCut) [inline src/invaders.rs]: There's no benefit to using a reference here. `char` is a `Copy` type, and it's just 32 bits. A reference creates a 64-bit pointer (or 32-bit if you have a 32-bit processor) to the location of 32 bits, requiring extra time to retrieve, extra space to store, and extra syntax to deal with.

```suggestion
    face1: char,
    face2: char,
```

File: src/terminal.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+        Ok(Terminal { stdout })
+    }
+}
+
+impl Drop for Terminal {
+    fn drop(&mut self) {
+        self.stdout.execute(Show).unwrap();
+        self.stdout.execute(LeaveAlternateScreen).unwrap();
+        terminal::disable_raw_mode().unwrap_or(());
```
Comment (CleanCut) [inline src/terminal.rs]: Rather than crash the program if part of the terminal cleanup fails, let's ignore any errors that occur so we can continue attempting to clean up.

```suggestion
        let _ = self.stdout.execute(Show);
        let _ = self.stdout.execute(LeaveAlternateScreen);
        let _ = terminal::disable_raw_mode();
```

File: src/terminal.rs [hunk 3]
```diff
@@ -0,0 +1,29 @@
+use crossterm::{
+    cursor::{Hide, Show},
+    terminal::{self, EnterAlternateScreen, LeaveAlternateScreen},
+    ExecutableCommand,
+};
+use std::io::{self, Stdout};
+
+pub struct Terminal {
+    stdout: Stdout,
+}
```
Comment (CleanCut) [inline src/terminal.rs:10]: I like your idea of encapsulating the terminal initialization/cleanup as a struct with a destructor!

File: src/invaders.rs [hunk 4]
```diff
@@ -21,48 +23,54 @@ pub struct Invaders {
...
-                    && (x < NUM_COLS - 2)
-                    && (y > 0)
-                    && (y < 9)
-                    && (x % 2 == 0)
-                    && (y % 2 == 0)
-                {
-                    army.push(Invader { x, y, points: 1 });
-                }
+        let mut faces_list = INVADER_FACES.iter();
+        for y in (INVADERS_FIRST_ROW..=INVADERS_LAST_ROW).step_by(INVADERS_ROW_INTERVAL) {
```
Comment (CleanCut) [inline src/invaders.rs]: Oooo, your idea to use `step_by` is quite nice!

I see that you're trying to iterate through the `y` values and the faces at the same rate, and resorted to manual iteration for the faces. Let's iterate through both at once with [zip](https://doc.rust-lang.org/std/iter/fn.zip.html)!

```suggestion
        for (y, faces) in (INVADERS_FIRST_ROW..=INVADERS_LAST_ROW)
            .step_by(INVADERS_ROW_INTERVAL)
            .zip(INVADER_FACES.iter())
        {
```

File: src/invaders.rs [hunk 5]
```diff
@@ -21,48 +23,54 @@ pub struct Invaders {
...
+            // unwrap is safe as the INVADER_FACES is guarded by INVADERS_ROW_COUNT
```
Comment (CleanCut) [inline src/invaders.rs:30]: ❤️ the unwrap comment, though it should go away due to the previous suggestion.

File: src/main.rs [hunk 6]
```diff
@@ -74,7 +71,7 @@ fn main() -> Result<(), Box<dyn Error>> {
 
         if in_menu {
             // Input handlers for the menu
-            while event::poll(Duration::default())? {
+            while event::poll(Duration::from_millis(KEYBOARD_POLLING_TIMEOUT_MS))? {
```
Comment (CleanCut) [inline src/main.rs]: What is the rationale behind this change?

`Duration::default()` is zero seconds, which was chosen because I didn't want to introduce any delay reading the keyboard events. There is a separate `sleep` at the end of the loop aimed at reducing CPU usage.

A better improvement would be to change this to `Duration::ZERO` to be more explicit.

**Review by CleanCut — CHANGES_REQUESTED** [review]
First off, thank you! ❤️  I appreciate the work you've put into this. I think it would be a benefit to create some constants.

Second, apologies for the delay. My life has been crazy this month.

Third, in addition to some specific suggestions I made inline, I think there are a few general improvements I would like to see before accepting this. If you're willing to make them, then great! If not, then I'll eventually find some time to do it myself.

---

General Improvements:

1. While I'm a fan of avoiding [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)) when it makes sense, I have nothing against literals in general. 
   - To me, the string constants feel less understandable than the strings themselves. Let's revert the strings to literals as they were before.
   - However, if you think the string+filename could be better described, I'd be fine changing the string values. For example, we could change `pew` to `shot` (with the corresponding file name change) if it were clearer.
   - Let's revert the array of strings to an in-place literal as well.

2) If a constant is only used in one file, it should live in that file (near the top, under the `use` statements) as a _private_ constant.

3) Let's try to reduce the amount of constants we have. In particular, I feel like some of the ones that are calculated from others would be better as logic inside of `Invaders::new()`

Also, not all numbers need to be constants. If a number is _only used once_, then it is sometimes clearer to leave it as a number. Ask yourself: Is it clear what this number means?  If the meaning is clear, and it's only used once, then a literal number is okay.

If the meaning isn't clear, but the number is only used once, then there's one more question to ask yourself: Would this be clearer with a comment or as a constant (or both)? It's more of an art than a science, so there's not a hard rule about which is better.

```
// The minimum interval of time allowed between invader movements, in milliseconds.
const MIN_MOVE_INTERVAL: u64 = 250;
```

4) Let's make these constants more concise! Three words max. Strive for one or two. `JAVA_AND_OBJECTIVE_C_PROMOTE_USING_FULL_SENTENCES_IN_THEIR_IDENTIFIERS_BUT_NOT_RUST`. 😆 🤣 

For example:

- `INVADERS_DIRECTION_LEFT`.  It is clear from that `LEFT` is a direction, so `DIRECTION` can be removed. The direction is not specific to invaders, so there's no need to prefix it with `INVADERS` (also, it is only used in `invaders.rs` so it should live there) That leaves us with just `LEFT`. Much shorter! Still clear.
- `INVADERS_MOVE_INTERVAL_MINIMAL_MS` -> `MIN_MOVE_INTERVAL`
- etc.

File: src/terminal.rs (hunk 3 above)
Comment (HaKr) [reply src/terminal.rs:10]: Thank you, It is one of those things I really like about the Rust way of working

File: src/invaders.rs (hunk 4 above)
Comment (HaKr) [reply src/invaders.rs]: I did read about Zip, but never gave it a thought here. But it is a very elegant and self-describing method

File: src/invaders.rs (hunk 5 above)
Comment (HaKr) [reply src/invaders.rs:30]: Very kind of you, but replacing a comment with a self-describing method is even more elegant

File: src/main.rs (hunk 6 above)
Comment (HaKr) [reply src/main.rs]: Actually, I was also looking at a way to get a predictable frames-per-second rate. You are right that using Duration::ZERO would help that even more, than forcing some timeout value.

I guess I sill feel a little uncomfortable with the Defaults::default() concept. Will take some time for me to get used to  ;-). Still I tend to lean towards the Duration::ZERO way of naming this

**Review by CleanCut — COMMENTED** [review]
Progress!

---

> 2) If a constant is only used in one file, it should live in that file (near the top, under the `use` statements) as a _private_ constant.

I think there's maybe just two or three constants that are used in more than one file, so let's move most of the constants to the files where they are used.

> 3) Let's try to reduce the amount of constants we have. In particular, I feel like some of the ones that are calculated from others would be better as logic inside of `Invaders::new()`

I'm thinking we can eliminate at least `INVADERS_LAST_ROW` and `INVADERS_FIRST_COL`

> 4) Let's make these constants more concise! Three words max. Strive for one or two.

Give it a try! ☝🏻

**Comment by CleanCut** [comment]
(the ci failure can be fixed by running `cargo fmt` and committing the changes)

**Comment by CleanCut** [comment]
Closing as this seems to have been abandoned. Feel free to reopen if you would like to pick the work back up.
